### ISSUE TITLE
Wrong rendering of Markdown

### ISSUE THREAD

**Issue by igormironchik** [doxygen/doxygen#8670]
**Describe the bug**

Markdown processor renders wrong the following code:

```
Why should I use this library?

 * There are tons of libraries for parsing command line arguments on the
Internet. But args-parser also provides the possibility to define commands, like

    ```
    git add -A
    git commit
    git push
    ```

    Help output is very user-friendly.

    If you need only simple arguments in style `--do-it <value>` then possibly
    you will find another library more useful for you, but who knows...

    This library is highly efficient in performance. On Intel (R) Core (TM)
    i5-7200U CPU @ 2.50GHz args-parser occupies 0.5 milliseconds for 1000
    arguments with value. Really, guys, do you look at performance or something
    that library can do? This library is very flexible with powerfull help
    printing. What do you need more?
```

**Expected behavior**

It should render as one paragraph in list item, like this:

Why should I use this library?

 * There are tons of libraries for parsing command line arguments on the
Internet. But args-parser also provides the possibility to define commands, like

    ```
    git add -A
    git commit
    git push
    ```

    Help output is very user-friendly.

    If you need only simple arguments in style `--do-it <value>` then possibly
    you will find another library more useful for you, but who knows...

    This library is highly efficient in performance. On Intel (R) Core (TM)
    i5-7200U CPU @ 2.50GHz args-parser occupies 0.5 milliseconds for 1000
    arguments with value. Really, guys, do you look at performance or something
    that library can do? This library is very flexible with powerfull help
    printing. What do you need more?

But it renders as:

Why should I use this library?

 * There are tons of libraries for parsing command line arguments on the
Internet. But args-parser also provides the possibility to define commands, like

```
    ```
    git add -A
    git commit
    git push
    ```

    Help output is very user-friendly.

    If you need only simple arguments in style `--do-it <value>` then possibly
    you will find another library more useful for you, but who knows...

    This library is highly efficient in performance. On Intel (R) Core (TM)
    i5-7200U CPU @ 2.50GHz args-parser occupies 0.5 milliseconds for 1000
    arguments with value. Really, guys, do you look at performance or something
    that library can do? This library is very flexible with powerfull help
    printing. What do you need more?
```

**Screenshots**

![Screenshot_20210714_192042](https://user-images.githubusercontent.com/11455533/125657411-2aa40404-afca-4a2c-a6b2-c76afe04c1ae.png)

**Version**

Doxygen version 1.8.17 on Kubuntu 20.04.2

**Comment by albert-github** [comment]
The main problem here is the fact that the line after the list start line (so the line starting with "Internet") starts at the beginning of the line. The subsequent lines all start indented, even with 4 spaces) and the 4 extra spaces and these 4 extra spaces indicate the begin of a code section. The result is already better:
![image](https://user-images.githubusercontent.com/5223533/125657633-eedf5968-c6c6-4ea5-ab75-67ce5068afe1.png)

By removing the empty line after the "Internet" line and lining the rest of the paragraph with the "Internet" word the result is:
![image](https://user-images.githubusercontent.com/5223533/125657920-93eae919-b7c7-4896-93df-3cbecdfec48d.png)

The handling of the Markdown input is a tricky business especially in combination with used spaces and lists

Note: please always (as no cut / paste problems can occur):  attach a, small, self contained example (source+configuration file in a tar or zip) that allows us to reproduce the problem? Please don't add external links as they might not be persistent. 


I think the problem is not a bug but usage.

**Comment by igormironchik** [comment]
You are wrong. Yes, in Doxygen the approach you described will work. But Doxygen is wrong here.

Markdown standard says that separate paragraph in the list should be 4 spaces intended with the first
paragraph.

Look

```
    Why should I use this library?

     * There are tons of libraries for parsing command line arguments on the
    Internet. But args-parser also provides the possibility to define commands, like
    ```
    git add -A
    git commit
    git push
    ```
    Help output is very user-friendly.
```

This will render in Doxygen as I expect, but will be wrong rendered everywhere else.

Look at rendering with GitHub:

Why should I use this library?

 * There are tons of libraries for parsing command line arguments on the
Internet. But args-parser also provides the possibility to define commands, like
```
git add -A
git commit
git push
```
Help output is very user-friendly.

What is not a:

Why should I use this library?

 * There are tons of libraries for parsing command line arguments on the
Internet. But args-parser also provides the possibility to define commands, like

    ```
    git add -A
    git commit
    git push
    ```

    Help output is very user-friendly.

**Comment by albert-github** [comment]
You wrote:
> Markdown standard

which markdown standard?

Doxygen's output is based on https://daringfireball.net/projects/markdown/and what is written in the doxygen manual.

**Comment by igormironchik** [comment]
> which markdown standard?

https://spec.commonmark.org/

**Comment by igormironchik** [comment]
Will you argue that GitHub don't know how to render Markdown? Try here in comments...

**Comment by doxygen** [comment]
> Will you argue that GitHub don't know how to render Markdown? Try here in comments...

I think the problem is more that the original markdown specification is not very precise, and thus leaves room for interpretation. As a result, many dialects exist that all behave slightly differently. As I see it https://spec.commonmark.org is a welcome attempt to improve the specification (with the risk of creating another dialect of course) and doxygen might not adhere to this spec (it wasn't there when markdown support was added). I think it is a good idea to try to adhere to this spec, but  we have to be careful as it can also break existing documentation.

**Comment by igormironchik** [comment]
I can live with this issue, I don't need to update documentation very frequently, so I can fix wrongs by hands. But when under list item follows paragraph with indent equal to indent to first word in the list item - it's continuation of the list item. This is very important on my sight, and I believe that you will fix it. Thanks in any case. Doxygen is a good software, that helps.

**Comment by doxygen** [comment]
I've just pushed 2 commits that should address the problem. 

I noticed parsing was also wrong even when markdown support was disabled and `@code` was used for the code block instead of the 3 backticks. Also, the output of example 54 (`@parblock`) was wrong. All of these issues are also fixed. 

Please let me know if still see issues.

**Comment by igormironchik** [comment]
Thank you. Done, fixed. Good job.

![Screenshot_20210720_214740](https://user-images.githubusercontent.com/11455533/126378968-89d14bf3-640d-4e50-83fa-d46fbbdf1b5f.png)

**Comment by doxygen** [comment]
Good to hear. I do see an issue when generating doxygen's own documentation. I'm looking into that.

**Comment by albert-github** [comment]
It looks like that not only the mentioned warnings cause problems, but in general when there is an empty line just before an indented code block.It looks like that in the later case the indentation is ignored.

What does GitHub do?
**input**
```
Test on GitHub:

    this is now indented

and not indented anymore
```

**result**
Test on GitHub:

    this is now indented

and not indented anymore

(the extra space that looks to be present before the `this is now indented` is due to the settings of the layout elements in GitHub i.e. padding etc.)


I see the warnings and the indentation problem also in the CGAL results (I didn't check Fossies but here it will show up as well).

**Comment by albert-github** [comment]
The handling of the markdown has been improved by means of 
```
Commit: 7df152b7e1c289fc29ddf356c9bdab546f81eacf [7df152b]
Date: Wednesday, July 21, 2021 7:52:27 PM
Regression: fixed issue with generating doxygen's own documentation.

Also improved handling of markdown quotation blocks inside lists.
```
though there are still some differences:

- **Old** (i.e. before all changes)

  ![image](https://user-images.githubusercontent.com/5223533/126608665-9bf26e52-e96f-473d-b87f-b083c09cb914.png)

- **New**

  ![image](https://user-images.githubusercontent.com/5223533/126608709-ff9f85fa-50c3-401e-89c9-4e76316161a1.png)

- **old**

  ![image](https://user-images.githubusercontent.com/5223533/126609075-2151e48b-7f91-4aa8-8680-eaec6f774cb6.png)

- **new**

  ![image](https://user-images.githubusercontent.com/5223533/126608998-070ad70b-7c49-4686-acbb-a1c58bd2df57.png)


I also saw problems with (I just skimmed through the markdown page):
- Code Block Indentation
- Lists Extensions

**Comment by albert-github** [comment]
Looks like the commit fixes this issue.

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.2.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
